File: android/app/src/main/java/com/example/clawix/android/core/BridgeRuntimeState.kt
package com.example.clawix.android.core

/**
 * Bootstrap state of the daemon driving the BridgeServer. Mirrors
 * Swift `BridgeRuntimeState`. Sent immediately after `authOk` and on
 * every subsequent transition.
 */
sealed class BridgeRuntimeState {
    data object Booting : BridgeRuntimeState()
    data class Syncing(val message: String? = null) : BridgeRuntimeState()
    data object Ready : BridgeRuntimeState()
    data class Error(val message: String) : BridgeRuntimeState()

    val wireTag: String
        get() = when (this) {
            Booting -> "booting"
            is Syncing -> "syncing"
            Ready -> "ready"
            is Error -> "error"
        }

    companion object {
        fun fromWire(state: String, message: String?): BridgeRuntimeState = when (state) {
            "booting" -> Booting
            "syncing" -> Syncing(message)
            "ready" -> Ready
            "error" -> Error(message ?: "Unknown error")
            else -> Error("Unknown state: $state")
        }
    }
}

File: windows/Clawix.Core/BridgeRuntimeState.cs
namespace Clawix.Core;

public abstract record BridgeRuntimeState
{
    public abstract string WireTag { get; }

    public virtual string? ErrorMessage => null;

    public sealed record Booting : BridgeRuntimeState
    {
        public override string WireTag => "booting";
    }

    public sealed record Syncing : BridgeRuntimeState
    {
        public override string WireTag => "syncing";
    }

    public sealed record Ready : BridgeRuntimeState
    {
        public override string WireTag => "ready";
    }

    public sealed record Error(string Message) : BridgeRuntimeState
    {
        public override string WireTag => "error";

        public override string? ErrorMessage => Message;
    }
}
